io/private/copcwriter/Processor.cpp

```diff
@@ -0,0 +1,186 @@
+* conditions are met:
+*
+*     * Redistributions of source code must retain the above copyright
+*       notice, this list of conditions and the following disclaimer.
+*     * Redistributions in binary form must reproduce the above copyright
+*       notice, this list of conditions and the following disclaimer in
+*       the documentation and/or other materials provided
+*       with the distribution.
+*     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
+*       names of its contributors may be used to endorse or promote
+*       products derived from this software without specific prior
+*       written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+* COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+* OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+* AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+* OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+* OF SUCH DAMAGE.
+****************************************************************************/
+
+
+#include <numeric>
+#include <random>
+
+#include <pdal/StageFactory.hpp>
+#include <io/BufferReader.hpp>
+
+#include <lazperf/writers.hpp>
+
+#include "GridKey.hpp"
+#include "Processor.hpp"
+#include "PyramidManager.hpp"
+
+namespace pdal
+{
+namespace copcwriter
+{
+
+Processor::Processor(PyramidManager& manager, const VoxelInfo& v, const BaseInfo& b) :
+    m_vi(v), b(b), m_manager(manager)
+{}
+
+
+void Processor::run()
+{
+    m_loader.init(b.pointFormatId, b.scaling, b.extraDims);
+    m_vi.initParentOctant();
+
+    size_t totalPoints = 0;
+    for (int i = 0; i < 8; ++i)
+    {
+        OctantInfo& child = m_vi.child(i);
+
+        totalPoints += child.numPoints();
+        if (child.numPoints() < MinimumPoints)
+            m_vi.octant().movePoints(child);
+    }
+    // It's possible that all the file infos have been moved above, but this is cheap.
+    if (totalPoints < MinimumTotalPoints)
+        for (int i = 0; i < 8; ++i)
+        {
+            OctantInfo& child = m_vi.child(i);
+            m_vi.octant().movePoints(child);
+        }
+
+    sample();
+    m_manager.queue(m_vi.octant());
+}
+
+
+void Processor::sample()
+{
+    // Accepted points are those that will go in this (the parent) cell.
+    // Rejected points will remain in the child cell they were in previously.
+
+    std::unique_ptr<std::mt19937> g;
+    if (b.opts.fixedSeed)
+    {
+        std::vector<int32_t> v{1234};
```

Comment:
Does this ever need to be a settable entity?

Reply:
No.